pkg/nas/controllerserver.go

```diff
@@ -102,97 +135,193 @@ func (cs *controllerServer) CreateVolume(ctx context.Context, req *csi.CreateVol
+			}
+		}
+
+		// step5: Mount nfs server to localpath
+		if err := DoMount(nfsServer, nfsPath, nfsVersion, nfsOptionsStr, mountPoint, req.Name); err != nil {
+			log.Errorf("CreateVolume: %s, Mount server: %s, nfsPath: %s, nfsVersion: %s, nfsOptions: %s, mountPoint: %s, with error: %s", req.Name, nfsServer, nfsPath, nfsVersion, nfsOptionsStr, mountPoint, err.Error())
+			return nil, errors.New("CreateVolume: " + req.Name + ", Mount server: " + nfsServer + ", nfsPath: " + nfsPath + ", nfsVersion: " + nfsVersion + ", nfsOptions: " + nfsOptionsStr + ", mountPoint: " + mountPoint + ", with error: " + err.Error())
+		}
+		if !CheckNfsPathMounted(mountPoint, nfsServer, nfsPath) {
+			return nil, errors.New("Check Mount nfsserver not mounted " + nfsServer)
+		}
+
+		// step6: create volume
+		fullPath := filepath.Join(mountPoint, pvName)
+		if err := os.MkdirAll(fullPath, 0777); err != nil {
+			log.Errorf("Provision: %s, creating path: %s, with error: %s", req.Name, fullPath, err.Error())
+			return nil, errors.New("Provision: " + req.Name + ", creating path: " + fullPath + ", with error: " + err.Error())
+		}
+		os.Chmod(fullPath, 0777)
+
+		// step7: Unmount nfs server
+		if !utils.Umount(mountPoint) {
+			log.Errorf("Provision: %s, unmount nfs mountpoint %s failed", req.Name, mountPoint)
+			return nil, errors.New("unable to unmount nfs server: " + nfsServer)
+		}
+
+		volumeContext["volumeAs"] = nasVol.VolumeAs
+		volumeContext["fileSystemId"] = ""
+		volumeContext["server"] = nfsServer
+		volumeContext["path"] = filepath.Join(nfsPath, pvName)
+		volumeContext["vers"] = nfsVersion
+		if nfsMode != "" {
+			volumeContext["mode"] = nfsMode
+		}
+		if value, ok := req.Parameters["options"]; ok && value != "" {
+			volumeContext["options"] = value
+		}
+
+		volSizeBytes := int64(req.GetCapacityRange().GetRequiredBytes())
+		tmpVol = &csi.Volume{
+			VolumeId:      req.Name,
+			CapacityBytes: int64(volSizeBytes),
+			VolumeContext: volumeContext,
+		}
 	}
+
 	pvcProcessSuccess[pvcUid] = true
 	log.Infof("Provision Successful: %s, with PV: %v", req.Name, tmpVol)
 	return &csi.CreateVolumeResponse{Volume: tmpVol}, nil
 }
 
-// call ecs api to delete disk
+// call nas api to delete disk
 func (cs *controllerServer) DeleteVolume(ctx context.Context, req *csi.DeleteVolumeRequest) (*csi.DeleteVolumeResponse, error) {
 	log.Infof("DeleteVolume: Starting deleting volume %s", req.GetVolumeId())
-	pvPath, nfsPath, nfsServer, nfsOptions := "", "", "", ""
 
 	pvInfo, err := cs.client.CoreV1().PersistentVolumes().Get(req.VolumeId, metav1.GetOptions{})
 	if err != nil {
 		return nil, fmt.Errorf("DeleteVolume: Get Volume: %s from cluster error: %s", req.VolumeId, err.Error())
 	}
+	volumeAs, fileSystemId, pvPath, nfsPath, nfsServer, nfsOptions := "", "", "", "", "", ""
+
+	if value, ok := pvInfo.Spec.CSI.VolumeAttributes["volumeAs"]; ok {
+		volumeAs = value
+	} else {
```

Comment:
not compatible with early version.

here can default as subPath. check the "path" of pv is empty, return error.